src/inlay-hints.ts

```diff
@@ -0,0 +1,233 @@
+// This file implements the client side of the proposed inlay hints
+// extension to LSP. The proposal is based on the one at
+// https://github.com/microsoft/language-server-protocol/issues/956,
+// with some modifications that reflect discussions in that issue.
+// The feature allows the server to provide the client with inline
+// annotations to display for e.g. parameter names at call sites.
+// The client-side implementation is adapted from rust-analyzer's.
+
+import * as vscode from 'vscode';
+import * as vscodelc from 'vscode-languageclient/node';
+
+import {ClangdContext, isClangdDocument} from './clangd-context';
+
+export function activate(context: ClangdContext) {
+  const feature = new InlayHintsFeature(context);
+  context.client.registerFeature(feature);
+}
+
+// Currently, only one hint kind (parameter hints) are supported,
+// but others (e.g. type hints) may be added in the future.
+enum InlayHintKind {
+  Parameter = 'parameter'
+}
+
+interface InlayHint {
+  range: vscodelc.Range;
+  kind: InlayHintKind;
+  label: string;
+}
+
+interface InlayHintsParams {
+  textDocument: vscodelc.TextDocumentIdentifier;
+}
+
+namespace InlayHintsRequest {
+export const type =
+    new vscodelc.RequestType<InlayHintsParams, InlayHint[], void>(
+        'clangd/inlayHints');
+}
+
+interface InlayDecorations {
+  // Hints are grouped based on whether they are positioned before or after
+  // the range, because different positions require different decoration types.
+  // The positioning is currently implied by the hint kind (e.g. parameter hints
+  // go before the range) rather than being explicit in the API.
+  beforeHints: vscode.DecorationOptions[];
+  afterHints: vscode.DecorationOptions[];
+}
+
+interface HintStyle {
+  decorationType: vscode.TextEditorDecorationType;
+
+  toDecoration(hint: InlayHint,
+               conv: vscodelc.Protocol2CodeConverter): vscode.DecorationOptions;
+}
+
+const beforeHintStyle = createHintStyle('before');
```

Comment:
I'd rather set up a map here from HintType -> HintStyle.
Otherwise we're adding afterHintStyle which is currently unused, and setting up an expectation that the only difference between kinds will be before vs after (I think we should do colors too, later)

Reply:
I went with a `parameterHintStyle` variable for now, to keep things simple.